server/src/main/java/org/elasticsearch/cluster/coordination/ElectionSchedulerFactory.java

```diff
@@ -0,0 +1,203 @@
+        TimeValue.timeValueMillis(100), TimeValue.timeValueMillis(1), Property.NodeScope);
+
+    public static final Setting<TimeValue> ELECTION_MAX_TIMEOUT_SETTING = Setting.timeSetting(ELECTION_MAX_TIMEOUT_SETTING_KEY,
+        TimeValue.timeValueSeconds(10), TimeValue.timeValueMillis(200), Property.NodeScope);
+
+    private final TimeValue initialTimeout;
+    private final TimeValue backoffTime;
+    private final TimeValue maxTimeout;
+    private final ThreadPool threadPool;
+    private final Random random;
+
+    public ElectionSchedulerFactory(Settings settings, Random random, ThreadPool threadPool) {
+        super(settings);
+
+        this.random = random;
+        this.threadPool = threadPool;
+
+        initialTimeout = ELECTION_INITIAL_TIMEOUT_SETTING.get(settings);
+        backoffTime = ELECTION_BACK_OFF_TIME_SETTING.get(settings);
+        maxTimeout = ELECTION_MAX_TIMEOUT_SETTING.get(settings);
+    }
+
+    /**
+     * Start the process to schedule repeated election attempts.
+     *
+     * @param gracePeriod       An initial period to wait before attempting the first election.
+     * @param scheduledRunnable The action to run each time an election should be attempted.
+     */
+    public Releasable startElectionScheduler(TimeValue gracePeriod, Runnable scheduledRunnable) {
+        final ElectionScheduler currentScheduler = new ElectionScheduler();
+        currentScheduler.scheduleNextElection(gracePeriod, scheduledRunnable);
+        return currentScheduler;
+    }
+
+    @SuppressForbidden(reason = "Argument to Math.abs() is definitely not Long.MIN_VALUE")
+    private static long nonNegative(long n) {
+        return n == Long.MIN_VALUE ? 0 : Math.abs(n);
+    }
+
+    /**
+     * @param randomSupplier supplier of randomly-chosen longs
+     * @param upperBound     exclusive upper bound
+     */
+    // package-private for testing
+    static long randomPositiveLongLessThan(LongSupplier randomSupplier, long upperBound) {
+        assert 1 < upperBound : upperBound;
+        return nonNegative(randomSupplier.getAsLong()) % (upperBound - 1) + 1;
+    }
+
+    @Override
+    public String toString() {
+        return "ElectionSchedulerFactory{" +
+            "initialTimeout=" + initialTimeout +
+            ", backoffTime=" + backoffTime +
+            ", maxTimeout=" + maxTimeout +
+            '}';
+    }
+
+    private class ElectionScheduler implements Releasable {
+        /**
+         * The current maximum timeout: the next election is scheduled randomly no later than this number of milliseconds in the future. On
+         * each election attempt this value is increased by `backoffTime`, up to the `maxTimeout`, to adapt to higher-than-expected latency.
+         */
+        private final AtomicLong currentMaxTimeoutMillis = new AtomicLong(Math.min(initialTimeout.millis(), maxTimeout.millis()));
```

Comment:
why not enforce the condition on maxTimeout to be >= initialTimeout?

Reply:
If `maxTimeout < initialTimeout` then nothing particularly bad happens. I added this call to `Math.min` because the spec is more uniform with it there than not.

Reply:
sure, nothing bad happens, but it might still be a user configuration error (user configured initalTimeout, but forgot to set maxTimeout). I think it's nicer to just throw an exception and report it.